nomad/volumewatcher/volume_watcher.go

```diff
@@ -0,0 +1,382 @@
+	vw.exitFn()
+}
+
+func (vw *volumeWatcher) isRunning() bool {
+	vw.wLock.RLock()
+	defer vw.wLock.RUnlock()
+	select {
+	case <-vw.shutdownCtx.Done():
+		return false
+	case <-vw.ctx.Done():
+		return false
+	default:
+		return vw.running
+	}
+}
+
+// watch is the long-running function that watches for changes to a volume.
+// Each pass steps the volume's claims through the various states of reaping
+// until the volume has no more claims eligible to be reaped.
+func (vw *volumeWatcher) watch() {
+	for {
+		select {
+		// TODO(tgross): currently server->client RPC have no cancellation
+		// context, so we can't stop the long-runner RPCs gracefully
+		case <-vw.shutdownCtx.Done():
+			return
+		case <-vw.ctx.Done():
+			return
+		case vol := <-vw.updateCh:
+			// while we won't make raft writes if we get a stale update,
+			// we can still fire extra CSI RPC calls if we don't check this
+			if vol == nil || vw.v == nil || vol.ModifyIndex >= vw.v.ModifyIndex {
+				vol = vw.getVolume(vol)
+				if vol == nil {
+					return
+				}
+				vw.volumeReap(vol)
+			}
+		}
+	}
+}
+
+// getVolume returns the tracked volume, fully populated with the current
+// state
+func (vw *volumeWatcher) getVolume(vol *structs.CSIVolume) *structs.CSIVolume {
+	vw.wLock.RLock()
+	defer vw.wLock.RUnlock()
+
+	var err error
+	ws := memdb.NewWatchSet()
+
+	vol, err = vw.state.CSIVolumeDenormalizePlugins(ws, vol.Copy())
+	if err != nil {
+		vw.logger.Error("could not query plugins for volume", "error", err)
+		return nil
+	}
+
+	vol, err = vw.state.CSIVolumeDenormalize(ws, vol)
+	if err != nil {
+		vw.logger.Error("could not query allocs for volume", "error", err)
+		return nil
+	}
+	vw.v = vol
+	return vol
+}
+
+// volumeReap collects errors for logging but doesn't return them
+// to the main loop.
+func (vw *volumeWatcher) volumeReap(vol *structs.CSIVolume) {
+	vw.logger.Trace("releasing unused volume claims", "id", vol.ID, "namespace", vol.Namespace)
+	err := vw.volumeReapImpl(vol)
+	if err != nil {
+		vw.logger.Error("error releasing volume claims", "error", err)
+	}
+	if vw.isUnclaimed(vol) {
+		vw.Stop()
+	}
+}
+
+func (vw *volumeWatcher) isUnclaimed(vol *structs.CSIVolume) bool {
+	return len(vol.ReadClaims) == 0 && len(vol.WriteClaims) == 0 && len(vol.PastClaims) == 0
+}
+
+func (vw *volumeWatcher) volumeReapImpl(vol *structs.CSIVolume) error {
+	var result *multierror.Error
+	nodeClaims := map[string]int{} // node IDs -> count
+	jobs := map[string]bool{}      // namespace+jobID -> stopped
+
+	// if a job is purged, the subsequent alloc updates can't
+	// trigger a GC job because there's no job for them to query.
+	// Job.Deregister will send a claim release on all claims
+	// but the allocs will not yet be terminated. save the status
+	// for each job so that we don't requery in this pass
+	checkStopped := func(jobID string) bool {
+		namespace := vw.v.Namespace
+		isStopped, ok := jobs[namespace+jobID]
```

Comment:
Isn't namespace in any given vw always going to be the same so that just `jobID` is unique here?

Reply:
Good catch. Will fix.